Document tt export/import (#3737)
tarantool/doc: doc/reference/tooling/tt_cli/import.rst

.. code-block:: console

    $ tt [crud] import URI FILE SPACE [IMPORT_OPTION ...]
    # or
    $ tt [crud] import URI - SPACE < FILE [IMPORT_OPTION ...]

``tt [crud] import`` imports data from a file to a space.
The ``crud`` command is optional and can be used to import data to a cluster by using the `CRUD <https://github.com/tarantool/crud>`_ module. Without ``crud``, data is imported using the :ref:`box.space <box_space>` API.

This command takes the following arguments:

*   ``URI``: The URI of a router instance if ``crud`` is used. Otherwise, it should specify the URI of a storage.
*   ``FILE``: The name of a file containing data to be imported.
*   ``SPACE``: The name of a space to which data is imported.

..  NOTE::

    You should have :ref:`write access <authentication-owners_privileges>` to the required space and `execute` access to `universe` to be able to import data.

..  NOTE::

    Importing isn't supported for the :ref:`interval <index-box_interval>` field type.

.. _tt-import-match-fields:

Matching of input and space fields
----------------------------------

.. _tt-import-match-fields-auto:

Automatic matching
~~~~~~~~~~~~~~~~~~

> Suppose, you have the ``customers.csv`` file with a header containing field names in the first row:

Reviewer: ```suggestion
Suppose that you have the ``customers.csv`` file with a header containing field names in the first row:
```

.. code-block:: text

    id,firstname,lastname,age
    1,Andrew,Fuller,38
    2,Michael,Suyama,46
    3,Robert,King,33
    # ...

If the target ``customers`` space has fields with the same names, you can import data using the ``--header`` and ``--match`` options specified as follows:

.. code-block:: console

    $ tt crud import localhost:3301 customers.csv customers \
                     --header \
                     --match=header

In this case, fields in the input file and the target space are matched automatically.
You can also match fields :ref:`manually <tt-import-match-fields-manual>` if field names in the input file and the target space differ.
Note that if you're importing data into a cluster, you don't need to specify the ``bucket_id`` field.
The CRUD module generates ``bucket_id`` values automatically.

.. _tt-import-match-fields-manual:

Manual matching
~~~~~~~~~~~~~~~

> The ``--match`` option enables you to import data by matching field names in the input file and the target space manually.

Reviewer: ```suggestion
The ``--match`` option enables importing data by matching field names in the input file and the target space manually.
```

                     --header \
                     --match "id=customer_id;firstname=name;lastname=surname;age=customer_age"

Similarly, you can configure mapping using numeric field positions in the input file:

.. code-block:: console

    $ tt crud import localhost:3301 customers.csv customers \
                     --header \
                     --match "id=1;firstname=2;lastname=3;age=4"

Below are the rules if some fields are missing in input data or space:

*   If a space has fields that are not specified in input data, ``tt [crud] import`` tries to insert ``null`` values.
*   If input data contains fields missing in a target space, these fields are ignored.

.. _tt-import-duplicate-error:

Handling duplicate primary key errors
-------------------------------------

The ``--on-exist`` option enables you to control data import when a duplicate primary key error occurs.
In the example below, values already existing in the space are replaced with new ones:

.. code-block:: console

    $ tt crud import localhost:3301 customers.csv customers \
                     --on-exist replace

.. _tt-import-parsing-error:

Handling parsing errors
-----------------------

> To skip rows whose data cannot be parsed correctly, you can use the ``--on-error`` option as follows:

Reviewer: ```suggestion
To skip rows whose data cannot be parsed correctly, use the ``--on-error`` option as follows:
```

..  option:: --password STRING

    Specify a password used to connect to the instance.

..  option:: --progress STRING

    Specify the name of a progress file that stores the following information:

    *   The positions of lines that were not imported at the last launch.
    *   The last position that was processed at the last launch.

    If a file with the specified name exists, it is taken into account when importing data.
    ``tt import`` tries to insert lines that were not imported and then continues importing from the last position.

    At each launch, the content of a progress file with the specified name is overwritten.
    If the file with the specified name does not exist, a progress file is created with the results of this run.

    .. NOTE::

        If the option is not set, then this mechanism is not used.

..  option:: --quote STRING

    Set the symbol that defines a quote.
    For CSV, double quotes are used by default (``"``).
    The double symbol of this option acts as the escaping symbol within input data.

..  option:: -success STRING

    The name of a file with rows that were imported (the default is ``success``).

>     Overrides the file if it already exists.

Reviewer: Can't understand what exactly this means. Overwrites?

Author: exactly
